src/attached_probe.cpp

```diff
@@ -329,46 +389,106 @@ void AttachedProbe::resolve_offset_uprobe(bool safe_mode)
+    "/usr/lib/debug/boot/vmlinux-%1$s",
+    "/usr/lib/debug/boot/vmlinux-%1$s.debug",
+    "/usr/lib/debug/lib/modules/%1$s/vmlinux",
+    nullptr,
+  };
+
+  for (int i = 0; vmlinux_locs[i]; i++)
+  {
+    char path[PATH_MAX + 1];
+    snprintf(path, PATH_MAX, vmlinux_locs[i], buf.release);
+    if (access(path, R_OK))
+      continue;
+    return path;
+  }
 
-  if (AlignState::Ok == aligned)
-    return;
+  return "";
+}
 
-  // If we did not allow unaligned uprobes in the
-  // compile time, force the safe mode now.
-#ifndef HAVE_UNSAFE_UPROBE
+void AttachedProbe::resolve_offset_kprobe(bool safe_mode)
+{
+  struct bcc_symbol_option option = {};
+  struct symbol sym = {};
+  std::string &symbol = probe_.attach_point;
+  uint64_t func_offset = probe_.func_offset;
+  offset_ = func_offset;
+
+#ifndef HAVE_UNSAFE_PROBE
   safe_mode = true;
 #endif
 
-  switch (aligned)
-  {
-    case AlignState::NotAlign:
-      if (safe_mode)
-        throw std::runtime_error("Could not add uprobe into middle of instruction: " + tmp);
-      else
-        std::cerr << "Unsafe uprobe in the middle of the instruction: " << tmp << std::endl;
-      break;
+  if (func_offset == 0)
+    return;
 
-     case AlignState::Fail:
-       if (safe_mode)
-         throw std::runtime_error("Failed to check if uprobe is in proper place: " + tmp);
-       else
-         std::cerr << "Unchecked uprobe: " << tmp << std::endl;
-       break;
+  sym.name = symbol;
+  option.use_debug_file = 0;
+  option.use_symbol_type = BCC_SYM_ALL_TYPES;
 
-     case AlignState::NotSupp:
-       if (safe_mode)
-         throw std::runtime_error("Can't check if uprobe is in proper place (compiled without uprobe offset support): " + tmp);
-       else
-         std::cerr << "Unchecked uprobe: " << tmp << std::endl;
-       break;
+  std::string path = find_vmlinux();
+  if (path.empty())
+  {
+    if (safe_mode)
+    {
+      std::stringstream buf;
+      buf << "Could not find vmlinux to check the offset.";
+      buf << " Use BPFTRACE_VMLINUX env variable to specify vmlinux path.";
+#ifdef HAVE_UNSAFE_PROBE
+      buf << " Use --unsafe to skip the userspace check.";
+#else
+      buf << " Compile bpftrace with ALLOW_UNAFE_PROBE option to force skip "
+             "the check.";
+#endif
+      throw std::runtime_error(buf.str());
+    }
+    else
+    {
+      // linux kernel checks alignment, but not the function bounds
+      if (bt_verbose)
+        std::cout << "No vmlinux found. Skip offset checking." << std::endl;
+      return;
+    }
+  }
+  if (bt_verbose)
+    std::cout << "vmlinux: using " << path << std::endl;
+  bcc_elf_foreach_sym(path.c_str(), sym_name_cb, &option, &sym);
 
-     default:
-       throw std::runtime_error("Internal error: " + tmp);
+  if (!sym.start)
+    throw std::runtime_error("Could not resolve symbol: " + path + ":" +
+                             symbol);
+
+  if (func_offset >= sym.size)
+  {
+    std::stringstream ss;
```

Comment:
Don't think you need a stringstream here. You could just `std::to_string(sym.size)` in the exception constructor.